src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1818,9 +1820,16 @@ public SocketError SendToAsync(Memory<byte> buffer, int offset, int count, Socke
 
             bytesSent = 0;
             SocketError errorCode;
-            int observedSequenceNumber;
-            if (_sendQueue.IsReady(this, out observedSequenceNumber) &&
-                SocketPal.TryCompleteSendTo(_socket, buffer.Span, ref offset, ref count, flags, socketAddress, socketAddressLen, ref bytesSent, out errorCode))
+            int observedSequenceNumber = 0;
+
+            // The default size of the OS send buffer is 16KB. Because of that, sending a low number of bytes is almost never blocking.
+            // For such scenarios, the lock taken in "IsReady" is relatively expensive compared to the send operation itself.
+            // To avoid the expensive lock for "fast sends", a lock-free speculative check is performed.
+            bool isReady = count <= 1024
+                    ? _sendQueue.IsReady_Speculative()
+                    : _sendQueue.IsReady(this, out observedSequenceNumber);
```

Comment:
If we change it for send, we can do it for all operations.

I guess we can remove the lock in IsReady completely at that point and name it IsReadySpeculatively.